youtube_transcript_api/_transcripts.py

```diff
@@ -0,0 +1,202 @@
+if sys.version_info.major == 2: # pragma: no cover
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
+
+import json
+
+from xml.etree import ElementTree
+
+import re
+
+from ._html_unescaping import unescape
+from ._errors import VideoUnavailable, NoTranscriptFound, TranscriptsDisabled
+from ._settings import WATCH_URL
+
+
+class TranscriptDataFetcher():
+    def __init__(self, http_client):
+        self._http_client = http_client
+
+    def fetch(self, video_id):
+        return TranscriptData.build(
+            self._http_client,
+            video_id,
+            self._extract_captions_json(self._fetch_html(video_id), video_id)
+        )
+
+    def _extract_captions_json(self, html, video_id):
+        splitted_html = html.split('"captions":')
+
+        if len(splitted_html) <= 1:
+            if '"playabilityStatus":' not in html:
+                raise VideoUnavailable(video_id)
+
+            raise TranscriptsDisabled(video_id)
+
+        return json.loads(splitted_html[1].split(',"videoDetails')[0].replace('\n', ''))[
+            'playerCaptionsTracklistRenderer'
+        ]
+
+    def _fetch_html(self, video_id):
+        return self._http_client.get(WATCH_URL.format(video_id=video_id)).text.replace(
+            '\\u0026', '&'
+        ).replace(
+            '\\', ''
+        )
+
+
+class TranscriptData():
+    # TODO implement iterator
+
+    def __init__(
+        self, http_client, video_id, manually_created_transcripts, generated_transcripts, translation_languages
+    ):
+        self._http_client = http_client
+        self.video_id = video_id
+        self._manually_created_transcripts = manually_created_transcripts
+        self._generated_transcripts = generated_transcripts
+        self._translation_languages = translation_languages
+
+    @staticmethod
+    def build(http_client, video_id, captions_json):
+        manually_created_transcripts = []
+        generated_transcripts = []
+
+        for caption in captions_json['captionTracks']:
+            (generated_transcripts if caption.get('kind', '') == 'asr' else generated_transcripts).append(
+                {
+                    'url': caption['baseUrl'],
+                    'language': caption['name']['simpleText'],
+                    'language_code': caption['languageCode'],
+                    'is_generated': caption.get('kind', '') == 'asr',
+                    'is_translatable': caption['isTranslatable'],
+                }
+            )
+
+        return TranscriptData(
+            http_client,
+            video_id,
+            manually_created_transcripts,
+            generated_transcripts,
+            [
+                {
+                    'language': translation_language['languageName']['simpleText'],
+                    'language_code': translation_language['languageCode'],
+                } for translation_language in captions_json['translationLanguages']
+            ],
+        )
+
+    def find_transcript(self, language_codes):
```

Comment:
The problem i see with this right now is there is a presumed bug where `manually_created_transcripts` is always empty, meaning all the transcripts are being searched through at once. I'm not sure, in this case, that you can guarantee a manual caption is returned before an asr. However, the system seems to be intended to run through all the `manually_created_transcripts` before the `generated` ones, which means that a language of lower priority but manual caption will return before a higher priority but asr caption track. 

I think a new structure with all the captions will probably be the best solution; maybe a dictionary where keys are language codes and their values are dictionaries like `{'manual' : { < caption info > }, 'asr': {< caption info> }}`?

Reply:
Yeah, that absolutely is a bug! Thank you for spotting!

Actually, now that I come to think about it, I made the internal structure list based instead of dict based to allow for fussy matching like `'en'` matching for `'en-GB'` (as you also pointed out), but then decided against it later on. I think I'll change it to dicts, as lookups are more efficient.